lib/AST/Decl.cpp

```diff
@@ -4087,46 +4087,152 @@ bool ProtocolDecl::isAvailableInExistential(const ValueDecl *decl) const {
   return true;
 }
 
-bool ProtocolDecl::existentialTypeSupportedSlow(LazyResolver *resolver) {
+void ProtocolDecl::getAssociatedTypeFreedomMapRec(
+                     AssociatedTypeFreedomMap &map) const {
+  for (auto *proto: getInheritedProtocols())
+    proto->getAssociatedTypeFreedomMapRec(map);
+
+  for (auto *assoc: getAssociatedTypeMembers())
+    map.insert({assoc, false});
+
+  for (auto &req : getRequirementSignature()) {
+    if (req.getKind() == RequirementKind::SameType &&
+        req.getSecondType()->getKind() != TypeKind::DependentMember) {
```

Comment:
This is not sufficient. The second type might be a generic type parameter type (eg, 'Self').